jnigen/lib/src/bindings/linker.dart

```diff
@@ -2,47 +2,95 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
-import 'visitor.dart';
 import '../config/config.dart';
 import '../elements/elements.dart';
+import '../logging/logging.dart';
+import 'visitor.dart';
 
 typedef _Resolver = ClassDecl Function(String? binaryName);
 
+/// Creates and links [ClassDecl] objects from imported dependencies.
 /// Adds references from child elements back to their parent elements.
 /// Resolves Kotlin specific `asyncReturnType` for methods.
-class Linker extends Visitor<Classes, void> {
+class Linker extends Visitor<Classes, Future<void>> {
   Linker(this.config);
 
   final Config config;
 
   @override
-  void visit(Classes node) {
-    final classLinker = _ClassLinker(config, (binaryName) {
-      return ClassDecl.predefined[binaryName] ??
+  Future<void> visit(Classes node) async {
+    // Specify paths for this package's classes.
+    final root = config.outputConfig.dartConfig.path;
+    if (config.outputConfig.dartConfig.structure ==
+        OutputStructure.singleFile) {
+      // Connect all to the root if the output is in single file mode.
+      final path = root.toFilePath();
+      for (final decl in node.decls.values) {
+        decl.path = path;
+      }
+    } else {
+      for (final decl in node.decls.values) {
+        final dollarSign = decl.binaryName.indexOf('\$');
+        final className = dollarSign != -1
+            ? decl.binaryName.substring(0, dollarSign)
+            : decl.binaryName;
+        final path = className.replaceAll('.', '/');
+        decl.path = root.resolve(path).toFilePath();
+      }
+    }
+
+    // Find all the imported classes.
+    await config.importClasses();
+
+    if (config.importedClasses.keys
+        .toSet()
+        .intersection(node.decls.keys.toSet())
+        .isNotEmpty) {
+      log.fatal(
+        'Trying to re-import the generated classes.\n'
+        'Try hiding the class(es) in import.',
+      );
+    }
+
+    for (final className in config.importedClasses.keys) {
+      log.finest('Imported $className successfully.');
+    }
+
+    ClassDecl resolve(String? binaryName) {
+      return config.importedClasses[binaryName] ??
           node.decls[binaryName] ??
-          ClassDecl.object;
-    });
+          resolve('java.lang.Object');
```

Comment:
Should we use something like `(TypeUsage.object.type as DeclaredType).binaryName` instead of repeating the string?
Should we have some special representation for Object? E.g. we default to it if we don't have a type. `ClassDecl.object`, `DeclaredType.object` etc.

Now that we have imports, maybe we shouldn't instantiate it manually (duplicating things between Dart/yaml), but just have a late final that does a `resolve` or something.

Reply:
As you can see in the red, I had the `ClassDecl.object` before. I think it's misplaced as a static field if we can realistically change it, maybe `config.object` is better. If we choose to never change `object`, then we can keep it hardcoded.

Reply:
Okay, I leave it to you to find the cleanest solution. 👌